### PR TITLE
Fix esdt cross sc fail revert

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

File: process/smartContract/process.go [hunk 1]
```diff
@@ -894,6 +902,12 @@ func (sc *scProcessor) isSCExecutionAfterBuiltInFunc(
 		return false, nil, nil
 	}
 
+	callType := determineCallType(tx)
+	if callType == vmcommon.AsynchronousCallBack {
```
Comment (SebastianMarian) [inline process/smartContract/process.go:906]: Is this a case of backwards compatibility?

File: process/smartContract/process.go [hunk 2]
```diff
@@ -903,10 +917,7 @@ func (sc *scProcessor) isSCExecutionAfterBuiltInFunc(
 	}
 
 	scExecuteOutTransfer := outAcc.OutputTransfers[0]
-	callType := determineCallType(tx)
-	txData := prependCallbackToTxDataIfAsyncCallBack(scExecuteOutTransfer.Data, callType)
-
-	function, arguments, err := sc.argsParser.ParseCallData(string(txData))
+	function, arguments, err := sc.argsParser.ParseCallData(string(scExecuteOutTransfer.Data))
```
Comment (SebastianMarian) [inline process/smartContract/process.go:920]: Is this a case of backwards compatibility?

File: process/smartContract/process.go [hunk 3]
```diff
@@ -942,23 +992,35 @@ func fillWithESDTValue(fullVMInput *vmcommon.ContractCallInput, newVMInput *vmco
...
 	if err != nil {
-		return false
+		return "", false
 	}
 
-	return function == core.BuiltInFunctionESDTTransfer
+	if function != core.BuiltInFunctionESDTTransfer {
+		return "", false
+	}
+	if len(args) < 2 {
```
Comment (SebastianMarian) [inline process/smartContract/process.go:1014]: Is this a case of backwards compatibility?

File: process/smartContract/process.go [hunk 4]
```diff
@@ -1422,8 +1484,9 @@ func (sc *scProcessor) createSCRsWhenError(
 	}
 
 	accumulatedSCRData := ""
-	if sc.isCrossShardESDTTransfer(tx) {
-		accumulatedSCRData += string(tx.GetData())
+	esdtReturnData, isCrossShardESDTCall := sc.isCrossShardESDTTransfer(tx)
+	if callType != vmcommon.AsynchronousCallBack && isCrossShardESDTCall {
```
Comment (SebastianMarian) [inline process/smartContract/process.go:1488]: Is this a case of backwards compatibility?

File: process/smartContract/process.go [hunk 5]
```diff
@@ -1657,17 +1720,25 @@ func (sc *scProcessor) useLastTransferAsAsyncCallBackWhenNeeded(
...
 		return false
 	}
 
 	_, err = sc.builtInFunctions.Get(function)
-	return err == nil
+	if err != nil {
+		return false
+	}
+
+	if function != core.BuiltInFunctionESDTTransfer {
```
Comment (SebastianMarian) [inline process/smartContract/process.go:1737]: Is this a case of backwards compatibility?

File: process/smartContract/process.go [hunk 6]
```diff
@@ -1657,17 +1720,25 @@ func (sc *scProcessor) useLastTransferAsAsyncCallBackWhenNeeded(
...
+		return true
+	}
+
+	return len(args) == 2
```
Comment (SebastianMarian) [inline process/smartContract/process.go:1741]: Is this a case of backwards compatibility?

File: process/smartContract/process.go [hunk 7]
```diff
@@ -732,7 +737,10 @@ func (sc *scProcessor) ExecuteBuiltInFunction(
 		return 0, err
 	}
 	if newVMOutput.ReturnCode != vmcommon.Ok {
-		return newVMOutput.ReturnCode, nil
+		if !check.IfNil(acntSnd) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:740]: yes it is. this code was not yet enabled. it gets enabled after built in function is enabled.

File: process/smartContract/process.go (hunk 1 above)
Comment (sasurobert) [reply process/smartContract/process.go:906]: yes it is. this code was not enabled yet.

File: process/smartContract/process.go (hunk 2 above)
Comment (sasurobert) [reply process/smartContract/process.go:920]: yes it is. this code was not enabled yet.

File: process/smartContract/process.go (hunk 3 above)
Comment (sasurobert) [reply process/smartContract/process.go:1014]: yes it is. this code was not enabled yet.

File: process/smartContract/process.go (hunk 4 above)
Comment (sasurobert) [reply process/smartContract/process.go:1488]: yes it is. this code was not enabled yet.

File: process/smartContract/process.go (hunk 5 above)
Comment (sasurobert) [reply process/smartContract/process.go:1737]: yes it is. this code was not enabled yet.

File: process/smartContract/process.go (hunk 6 above)
Comment (sasurobert) [reply process/smartContract/process.go:1741]: yes it is. this code was not enabled yet.

File: process/smartContract/process.go [hunk 8]
```diff
@@ -933,6 +944,45 @@ func (sc *scProcessor) isSCExecutionAfterBuiltInFunc(
 	return true, newVMInput, nil
 }
 
+func (sc *scProcessor) createVMInputWithAsyncCallBack(vmInput *vmcommon.ContractCallInput, vmOutput *vmcommon.VMOutput) *vmcommon.ContractCallInput {
```
Comment (iulianpascalau) [inline process/smartContract/process.go:947]: Maybe we should add some unit tests on the new code

**Review by iulianpascalau — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**